Various docs improvements from JOSS review

jagoosw opened the pull request:
This PR addresses several docs problems from the [JOSS review](https://github.com/openjournals/joss-reviews/issues/5669)

This includes

- [x] Adds model equations for all model components
- [x] Adds parameter variable name lists to each model component
- [x] Speeds up quickstart example by reducing resolution (and makes the example more realistic)
- [x] Adds flow diagrams explaining the bgc model formulations
- [x] Adds contributors guide (closes #128)
- [ ] Adds docs section explaining how to implement new models (closes #129)

jagoosw commented:
I can't work out how to write a piecewise equation nicely in Documenters latex. In normal latex you can write `\left\{ \begin{array}...\end{array}\right` but this doesn't work in Documenter

navidcy commented:
> I can't work out how to write a piecewise equation nicely in Documenters latex. In normal latex you can write `\left\{ \begin{array}...\end{array}\right` but this doesn't work in Documenter

Try escaping \ chars, eg

\\begin{array}…

jagoosw commented:
> > I can't work out how to write a piecewise equation nicely in Documenters latex. In normal latex you can write `\left\{ \begin{array}...\end{array}\right` but this doesn't work in Documenter
> 
> Try escaping \ chars, eg
> 
> ```
> \\begin{array}…
> ```

Think I've got this working now

navidcy reviewed (changes requested):
There is something wrong w rendering at

https://oceanbiome.github.io/OceanBioME.jl/previews/PR126/model_components/air-sea-gas/

jagoosw commented:
> There is something wrong w rendering at
> 
> https://oceanbiome.github.io/OceanBioME.jl/previews/PR126/model_components/air-sea-gas/

Good spot, thank you

navidcy commented:
Can we try size = (160, 32) for the first example? Movie looks bit pixelated now.

navidcy commented on README.md:
We also need to update the video link to the one made with the updated resolution.

jagoosw replied:
We could probably update it to show the one generated in the docs now instead of being a manual upload

jagoosw replied:
I had a think about this and don't think there's a way for us to get it from the correct branch/preview docs. Do you think we should change it to always get the video from the stable docs? This would only be an issue if we change the video and want to preview it

navidcy replied:
Yes! From the stable is better. Or the dev version of docs

jagoosw commented:
> Can we try size = (160, 32) for the first example? Movie looks bit pixelated now.

Sure, seems reasonable

jagoosw commented:
Todo:
- [ ] Add lists of parameter variable names to equations listings
- [ ] Add flow diagrams

navidcy reviewed (approved):
lgtm. I admit I didn’t check the equations (eg for sign errors and what not). 

I saw a todo list. If you want me to have a look at some point ping me.

jagoosw commented:
I've added tables to each page with equations listing the corresponding variable name for each parameter and added flow diagrams as per the review suggestions (will see how they look).

I also listed the units which we didn't have anywhere before.

jagoosw commented:
@navidcy I think I've finished making the changes here now. The README image doesn't appear to be loading from the preview which is a bit odd. I set it to the preview branch to test (intending to change to main once it worked) but it's not working.

navidcy commented on docs/src/index.md:
````diff
@@ -90,14 +90,14 @@ hm2 = heatmap!(ax2, xP, zP, Pₙ, colorrange = P_lims, colormap = Reverse(:bamak
 ```
 
-![buoyancy_front](buoyancy_front.mp4)
+![buoyancy_front](buoyancy_front.gif)
````
Just note that file size is bigger and this will have an imprint of the github repo size

jagoosw replied:
I'll change this back, I don't think its possible to have videos in the readme so thought this might work to fix it, but doesn't seem to.

navidcy commented:
> @navidcy I think I've finished making the changes here now. The README image doesn't appear to be loading from the preview which is a bit odd. I set it to the preview branch to test (intending to change to main once it worked) but it's not working.

Revert back to previous way of showing the video and leave this for a future PR? Upload the latest version of the gif in an issue and update the link?

navidcy commented on README.md:
```diff
@@ -28,13 +28,15 @@ julia> Pkg.add("OceanBioME")
 
 ## Running your first model
-As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front:
+As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front. This also requires Oceananigans, so we will add that first:
```
```suggestion
As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front. This example requires Oceananigans, so we install that first:
```

jagoosw commented:
I've changed to a `SeawaterBuoyancy` with constant salinity to reduce the confusion of also having b. It takes longer to run now though as I had to make the timestep smaller, we could reduce the buoyancy gradient which would probably make it so the timestep can be bigger?

navidcy commented on docs/src/index.md:
```diff
@@ -25,54 +25,62 @@ julia> Pkg.add("OceanBioME")
 ## Running your first model
 
-As a simple example let's run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front:
+As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front. This also requires Oceananigans, so we will add that first:
```
```suggestion
As a simple example lets run a Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model in a two-dimensional simulation of a buoyancy front. This example requires Oceananigans, so we install that first:
```

jagoosw commented:
Just tried changing the temp gradient and it looks fine still
![buoyancy_front](https://github.com/OceanBioME/OceanBioME.jl/assets/26657828/02a8aba3-0ee2-4f6b-b847-fe08fa600b7a)
So will change this and reduce the time step again

navidcy commented on docs/src/index.md:
````diff
@@ -25,54 +25,62 @@ julia> Pkg.add("OceanBioME")
-```@example quickstart
-using OceanBioME
-using Oceananigans, Oceananigans.Units
+```julia
````
Why did you change @example to simple code? That won’t work. Documenter wont run the code but rather will just print it.

jagoosw replied:
Ah I didn't mean todo that

navidcy replied:
(There is a chance I missed something here?)

jagoosw replied:
Think its because I copied from readme

navidcy replied:
Cool. Okie!

navidcy commented:
> Just tried changing the temp gradient and it looks fine still
> ![buoyancy_front](https://user-images.githubusercontent.com/26657828/259074376-02a8aba3-0ee2-4f6b-b847-fe08fa600b7a.gif)
> So will change this and reduce the time step again

Let’s not use a divergent colormap for T

jagoosw commented:
![buoyancy_front](https://github.com/OceanBioME/OceanBioME.jl/assets/26657828/ace41d0c-e213-45c5-9130-27c22b343626)

navidcy commented:
> Happy with this one?
> ![buoyancy_front](https://user-images.githubusercontent.com/26657828/259077208-ace41d0c-e213-45c5-9130-27c22b343626.gif)

Magnificent

jagoosw commented:
![buoyancy_front](https://github.com/OceanBioME/OceanBioME.jl/assets/26657828/6d39b9e2-2b44-4b79-814e-aec9703c87f0)

Maybe it's less pretty but a much larger domain like this might be a more realistic first example? And it runs in about the same time.

Obviously not really a realistic problem but the phytoplankton growth has a chance to develop and be effected by the mixing here?

jagoosw commented:
![buoyancy_front](https://github.com/OceanBioME/OceanBioME.jl/assets/26657828/61601698-9389-4e7a-be90-cd92cfdc9c36)
Improved (?) version with interpolation

jagoosw commented:
Or maybe something like this where the cold water is nutrient-rich, kind of in the vein of this paper https://www.math.hkust.edu.hk/~magan/AMCE6082010/paperpool2010/Franks_1992.pdf

![buoyancy_front](https://github.com/OceanBioME/OceanBioME.jl/assets/26657828/2b7c858e-3a93-4099-8747-291bd3766f31)

iuryt commented:
> Or maybe something like this where the cold water is nutrient-rich, kind of in the vein of this paper https://www.math.hkust.edu.hk/~magan/AMCE6082010/paperpool2010/Franks_1992.pdf ![buoyancy_front](https://user-images.githubusercontent.com/26657828/259180616-f9bc8e04-c605-4bd7-abb9-10d99f98f09e.gif)

It is very cool to see the effect of the downwelling at the front bringing phytoplankton to greater depths.
I like this version, also because you also show the evolution of the nutrients.

jagoosw commented:
> A thought: This PR has become large. Why not open new PR for contribution guide etc?

I've already done the contribution guide but forgot to push, so think I will add that and then do the other bit in a new PR?

jagoosw commented:
Happy for this PR to be merged now?

navidcy commented:
Merge!
